### PR TITLE
added support in metaManager for services and endpoints CRUD and query support in controller

### PR DESCRIPTION
**Pull request by sids-b**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds support in metaManager to receive service and endpoints updates from cloud. It also adds query support in controller. It is required for edgemesh development

**Which issue(s) this PR fixes**:
XREF: #673

### CODE REVIEW

File: edge/pkg/metamanager/msg_processor.go [hunk 1]
```diff
@@ -105,7 +115,7 @@ func parseResource(resource string) (string, string, string) {
 
 // is resource type require remote query
 func requireRemoteQuery(resType string) bool {
-	return resType == model.ResourceTypeConfigmap || resType == model.ResourceTypeSecret
+	return resType == model.ResourceTypeConfigmap || resType == model.ResourceTypeSecret || resType == constants.ResourceTypeService || resType == constants.ResourceTypeEndpoints
```
Comment (qizha) [inline edge/pkg/metamanager/msg_processor.go]: why we need edgemesh answer the remote query for those two resources?

File: edge/pkg/metamanager/msg_processor.go [hunk 2]
```diff
@@ -331,7 +428,11 @@ func (m *metaManager) processRemoteQuery(message model.Message) {
 			log.LOGGER.Errorf("update meta failed, %s", msgDebugInfo(&resp))
 		}
 		resp.BuildHeader(resp.GetID(), originalID, resp.GetTimestamp())
-		send2Edged(&resp, message.IsSync(), m.context)
+		if resType == constants.ResourceTypeService || resType == constants.ResourceTypeEndpoints {
```
[the remark this answers is not in the document]
Comment (sids-b) [reply edge/pkg/metamanager/msg_processor.go]: Suppose an endpoint or service is not found in edge, they can be queried from cloud.

File: edge/pkg/metamanager/msg_processor.go (hunk 1 above)
Comment (sids-b) [reply edge/pkg/metamanager/msg_processor.go]: Suppose an endpoint or service is not found in edge, they can be queried from cloud

File: edge/pkg/metamanager/msg_processor.go [hunk 3]
```diff
@@ -176,6 +195,69 @@ func (m *metaManager) processUpdate(message model.Message) {
 	}
 
 	resKey, resType, _ := parseResource(message.GetResource())
+	if resType == constants.ResourceTypeServiceList || resType == constants.ResourceTypeEndpointsList {
+		switch resType {
+		case constants.ResourceTypeEndpointsList:
+			var epsList []v1.Endpoints
+			err = json.Unmarshal(content, &epsList)
```
[the remark this answers is not in the document]
Comment (sids-b) [reply edge/pkg/metamanager/msg_processor.go]: We are sending all the endpoints to all services here. So unmarshalling and saving individual endpoints object in db for easy retrieval.

**Comment by sids-b** [comment]
> please get these codes checked in to edgemesh-dev branch. Thanks

Sure @qizha , I will raise this same PR to edgemesh branch after this PR is merged. This will keep the commit id same in both the edgemesh and master branch

File: edge/pkg/metamanager/msg_processor.go [hunk 4]
```diff
@@ -291,7 +384,11 @@ func (m *metaManager) processQuery(message model.Message) {
 	} else {
 		resp := message.NewRespByMessage(&message, *metas)
 		resp.SetRoute(MetaManagerModuleName, resp.GetGroup())
-		send2Edged(resp, message.IsSync(), m.context)
+		if resType == constants.ResourceTypeService || resType == constants.ResourceTypeEndpoints {
+			send2EdgeMesh(resp, message.IsSync(), m.context)
```
[the remark this answers is not in the document]
Comment (sids-b) [reply edge/pkg/metamanager/msg_processor.go]: Done, Thanks.

File: edge/pkg/metamanager/msg_processor.go [hunk 5]
```diff
@@ -154,6 +164,15 @@ func (m *metaManager) processInsert(message model.Message) {
 		feedbackError(err, "Error to save meta to DB", message, m.context)
 		return
 	}
+
+	if resType == constants.ResourceTypeService || resType == constants.ResourceTypeEndpoints {
```
[the remark this answers is not in the document]
Comment (sids-b) [reply edge/pkg/metamanager/msg_processor.go]: Done, Thanks

File: cloud/pkg/controller/controller/downstream.go [hunk 6]
```diff
@@ -212,13 +213,33 @@ func (dc *DownstreamController) syncEdgeNodes(stop chan struct{}) {
...
-							msg.Content = dc.lc.GetAllServices()
+							svcs := dc.lc.GetAllServices()
+							msg.Content = svcs
 							if err := dc.messageLayer.Send(*msg); err != nil {
 								log.LOGGER.Warnf("Send message failed with error: %s, operation: %s, resource: %s", err, msg.GetOperation(), msg.GetResource())
 							} else {
 								log.LOGGER.Infof("Send message successfully, operation: %s, resource: %s", msg.GetOperation(), msg.GetResource())
 							}
 
+							for _, svc := range svcs {
```
Comment (anyushun) [inline cloud/pkg/controller/controller/downstream.go]: This function has too many lines, split this function into multiple sub-functions according to the function please.

File: cloud/pkg/controller/controller/downstream.go [hunk 7]
```diff
@@ -319,44 +340,83 @@ func (dc *DownstreamController) syncEndpoints(stop chan struct{}) {
...
-			dc.lc.EdgeNodes.Range(func(key interface{}, value interface{}) bool {
-				nodeName, ok := key.(string)
-				if !ok {
-					log.LOGGER.Warnf("Failed to assert key to sting")
-					return true
+			if ok {
+				var listOptions metav1.ListOptions
+				var pods *v1.PodList
+				var err error
+				svc, ok := dc.lc.GetService(fmt.Sprintf("%s/%s", eps.Namespace, eps.Name))
```
Comment (anyushun) [inline cloud/pkg/controller/controller/downstream.go]: name in this function is ```${namespace}/${name}```， ```${name}/${namespace}``` in function GetServicePods, is this correct

File: cloud/pkg/controller/controller/downstream.go [hunk 8]
```diff
@@ -319,44 +340,83 @@ func (dc *DownstreamController) syncEndpoints(stop chan struct{}) {
...
+				if ok {
```
Comment (anyushun) [inline cloud/pkg/controller/controller/downstream.go]: too many lines in one function

File: cloud/pkg/controller/manager/location.go [hunk 9]
```diff
@@ -167,14 +173,65 @@ func (lc *LocationCache) GetAllServices() []v1.Service {
...
+func (lc *LocationCache) GetServicePods(name string) ([]v1.Pod, bool) {
+	value, ok := lc.servicePods.Load(name)
+	if !ok {
+		return []v1.Pod{}, ok
+	}
+	pods, ok := value.([]v1.Pod)
+	return pods, ok
+}
+
+// AddOrUpdateEndpoints in cache
```
Comment (anyushun) [inline cloud/pkg/controller/manager/location.go]: when endpoint updated, maybe some pod remove some label, these pod should be removed from lc.servicePods

**Comment by anyushun** [comment]
My concern: without considering edged, it's ok for metaManager to save podlist in sqlite and send to EdgeMesh, but what if edged query the podlist from metaManager when it starts, will edged run all the pods then? Please check this situation.

**Comment by sids-b** [comment]
> My concern: without considering edged, it's ok for metaManager to save podlist in sqlite and send to EdgeMesh, but what if edged query the podlist from metaManager when it starts, will edged run all the pods then? Please check this situation.

This situation is handled here. The pods which are to be added by edged are sent as individual pods message as well and not the podList. So the type is different. When you query resource type pod you will get the individual pod which are to be started by edged.

File: cloud/pkg/controller/manager/location.go (hunk 9 above)
Comment (sids-b) [reply cloud/pkg/controller/manager/location.go]: Nice point ! Yess this is taken care here https://github.com/kubeedge/kubeedge/pull/702/files#diff-55277caec6421504e2848e75c588c384R380

File: cloud/pkg/controller/controller/downstream.go (hunk 8 above)
Comment (sids-b) [reply cloud/pkg/controller/controller/downstream.go]: I will split later, Thanks

File: cloud/pkg/controller/controller/downstream.go (hunk 7 above)
Comment (sids-b) [reply cloud/pkg/controller/controller/downstream.go]: Nice Catch, modified to name/namespace thanks

Comment (sids-b) [reply cloud/pkg/controller/controller/downstream.go]: Sorry, modified to namespace/name

**Pull request merged**